### PR TITLE
[WIP]Merge different gpu backends with accelerator='gpu'

### LINKED ISSUE #13102: MPS (Mac M1) device support
## 🚀 Feature

https://pytorch.org/blog/introducing-accelerated-pytorch-training-on-mac/
Docs: https://pytorch.org/docs/master/notes/mps.html

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @akihironitta @rohitgr7 @justusschock

### PR DESCRIPTION
**Pull request by justusschock**
### THIS HAS TO BE MERGED AFTER #13636 

Setting `accelerator='gpu'` does no longer select the `GPUAccelerator` (see #13636 ), but will dynamically choose an available GPU backend (currently cuda or mps)

Fixes #13102 

### Does your PR introduce any breaking changes? If yes, please list them.
Yes, together with #13636 this changes `accelerator='gpu'` to dynamically select gpu backends

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

TODOs:
- [ ] Update Changelog
- [ ] Verify tests are passing

### CODE REVIEW

File: src/pytorch_lightning/utilities/enums.py [hunk 1]
```diff
@@ -242,15 +242,15 @@ class _AcceleratorType(LightningEnum):
...
-    >>> _AcceleratorType.GPU == 'GPU'
+    >>> _AcceleratorType.CUDA == 'CUDA'
     True
     >>> # which is case invariant
     >>> _AcceleratorType.TPU in ('tpu', 'CPU')
     True
     """
 
     CPU = "CPU"
-    GPU = "GPU"
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/enums.py:253]: Same comment about keeping GPU

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/lite/lite.py [hunk 2]
```diff
@@ -457,7 +458,7 @@ def _check_strategy_support(self, strategy: Optional[Union[str, Strategy]]) -> N
     def _supported_device_types() -> Sequence[_AcceleratorType]:
         return (
             _AcceleratorType.CPU,
-            _AcceleratorType.GPU,
```
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/lite/lite.py:460]: Actually no, because there is no GPUAccelerator (only the deprecated one replaced by the CUDAAccelerator) anymore, so we don't have a AcceleratorType anymore for this

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 3]
```diff
@@ -496,10 +501,20 @@ def _choose_accelerator(self) -> str:
...
 
+    @staticmethod
+    def _choose_gpu_accelerator_backend() -> str:
+        if CUDAAccelerator.is_available():
+            return "cuda"
+
+        if MPSAccelerator.is_available():
+            return "mps"
+
+        raise MisconfigurationException("No supported gpu backend found!")
```
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:516]: I thought about that, but I think this is consistent to the error type we have at https://github.com/Lightning-AI/lightning/blob/master/src/pytorch_lightning/trainer/connectors/accelerator_connector.py#L514

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -496,10 +501,20 @@ def _choose_accelerator(self) -> str:
                 return "hpu"
             if MPSAccelerator.is_available():
                 return "mps"
-            if torch.cuda.is_available() and torch.cuda.device_count() > 0:
+            if CUDAAccelerator.is_available():
                 return "cuda"
         return "cpu"
 
+    @staticmethod
+    def _choose_gpu_accelerator_backend() -> str:
```
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:509]: thought about that as well, but didn't go for it since this again implies the existence and usage of a gpu accelerator

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 5]
```diff
@@ -496,10 +501,20 @@ def _choose_accelerator(self) -> str:
...
+        if CUDAAccelerator.is_available():
+            return "cuda"
+
+        if MPSAccelerator.is_available():
```
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:513]: Will do that first thing tomorrow :)

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 6]
```diff
@@ -507,11 +507,10 @@ def _choose_auto_accelerator(self) -> str:
 
     @staticmethod
     def _choose_gpu_accelerator_backend() -> str:
-        if CUDAAccelerator.is_available():
-            return "cuda"
-
         if MPSAccelerator.is_available():
             return "mps"
+        if CUDAAccelerator.is_available():
+            return "cuda"
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:513]: noob q:

what's the reason for changing the order?

Comment (justusschock) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:513]: requested by @carmocca [here](https://github.com/Lightning-AI/lightning/pull/13642#discussion_r926964132)

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 7]
```diff
@@ -186,6 +186,10 @@ def __init__(
         self._amp_level_flag: Optional[str] = amp_level
         self._auto_select_gpus: bool = auto_select_gpus
 
+        # handle "gpu"
+        if accelerator == "gpu":
+            accelerator = self._choose_gpu_accelerator_backend()
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:191]: Why do we introduce a special case handling here?
I believe this should go into the `self._choose_accelerator()` as this falls under the "auto" case. 
I would prefer to follow the high-level structure we have established in the connector to stay organized.

Comment (justusschock) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:191]: The issue with that is that the accelerator-type is checked in `_check_config_and_set_final_flags` which is called before the actual auto-handling. This means that we still would require a special-casing there, which I tried to avoid by calling it earlier.

That was the main reason of https://github.com/Lightning-AI/lightning/pull/13642/commits/708b4b47c733a66fbd5d77abf4a7e7945e57d63f

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:514]: Which user input/hardware configuration would lead to this exception?
Let's add a corresponding test?

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:514]: having a cpu-only machine and requesting cpus. This was already covered by an old-test before moving. Need to check again now

**Review by akihironitta — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:514]: added the missing test

**Comment by carmocca** [comment]
@justusschock CI failures look real

**Comment by awaelchli** [comment]
Fixed it. After the changes we did for ddp fork #13405, we should mock the device parser functions instead of `torch.cuda.device_count` directly.
